### PR TITLE
implement dog plugin

### PR DESCRIPTION
**Pull request by BenTheElder**
Adds a random good doggo from a hooman curated list at https://random.dog/ when a comment contains `/(woof|bark)` :dog:

### CODE REVIEW

**Comment by ixdy** [comment]
that is a curious cat.

File: prow/plugins/dog/dog.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+var (
+	match = regexp.MustCompile(`(?mi)^/woof\s*$`)
```
Comment (fejta) [inline prow/plugins/dog/dog.go]: Also `/bark`?

File: prow/plugins/dog/dog.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+var client = http.Client{}
+
+type dogResult struct {
+	URL string `json:"url"`
+}
+
+func (dr dogResult) Format() (string, error) {
+	if dr.URL == "" {
+		return "", errors.New("empty source_url")
```
Comment (fejta) [inline prow/plugins/dog/dog.go]: Technically isn't a source_url key

File: prow/plugins/dog/dog.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		Examples:    []string{"/woof"},
+	})
+	return pluginHelp, nil
+}
+
+type githubClient interface {
+	CreateComment(owner, repo string, number int, comment string) error
+}
+
+type clowder interface {
```
Comment (fejta) [inline prow/plugins/dog/dog.go]: A pack of dogs, not a clowder

File: prow/plugins/dog/dog_test.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	"strings"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/github/fakegithub"
+)
+
+type fakeClowder string
```
Comment (fejta) [inline prow/plugins/dog/dog_test.go]: fakePack

File: prow/plugins/dog/dog_test.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+var human = flag.Bool("human", false, "Enable to run additional manual tests")
+var category = flag.String("category", "", "Request a particular category if set")
```
Comment (fejta) [inline prow/plugins/dog/dog_test.go]: Category is NA

File: prow/plugins/dog/dog_test.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+func (c fakeClowder) readDog() (string, error) {
+	return string(c), nil
+}
+
+func TestRealCat(t *testing.T) {
```
Comment (fejta) [inline prow/plugins/dog/dog_test.go]: not a cat

File: prow/plugins/dog/dog_test.go [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+		t.Skip("Real dogs disabled for automation. Manual users can add --human [--category=foo]")
+	}
+	if cat, err := dogURL.readDog(); err != nil {
+		t.Errorf("Could not read dog from %s: %v", dogURL, err)
+	} else {
+		fmt.Println(cat)
+	}
+}
+
+func TestFormat(t *testing.T) {
```
Comment (fejta) [inline prow/plugins/dog/dog_test.go]: I do not see where you are validating the behavior when the url returns unexpected json

**Comment by fejta** [comment]
/lgtm cancel
/hold cancel

Woops sorry, non trivial fixes needed :)

**Comment by ixdy** [comment]
seems like a lot of duplicated code in these plugins. maybe we need to refactor into a proper pet set?

File: prow/plugins/dog/dog_test.go (hunk 7 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog_test.go]: done now in `TestHttpResponse`

File: prow/plugins/dog/dog_test.go (hunk 6 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog_test.go]: done

File: prow/plugins/dog/dog_test.go (hunk 5 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog_test.go]: done

File: prow/plugins/dog/dog_test.go (hunk 4 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog_test.go]: done

File: prow/plugins/dog/dog.go (hunk 2 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog.go]: done

File: prow/plugins/dog/dog.go (hunk 3 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog.go]: done, somehow I thought this was some meme talk related to using cloud APIs :shrug:

File: prow/plugins/dog/dog.go (hunk 1 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog.go]: done, also updated usage and test cases to include this

**Comment by BenTheElder** [comment]
> We should consider combining yuks/cat/dog into a single fun plugin

We definitely could, but I think these serve as great examples of very simple plugins at the moment. Probably at least one of them should stay very small and self-contained.

**Comment by krzyzacy** [comment]
+1 to combine yuks/cat/dog so we can later on add many more :-)

heart could serve as a sample plugin for prow

File: prow/plugins/dog/dog_test.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+		shouldComment bool
+		shouldError   bool
+	}{
+		{
+			name:          "ignore edited comment",
+			state:         "open",
+			action:        github.GenericCommentActionEdited,
+			body:          "/woof",
+			shouldComment: false,
+			shouldError:   false,
```
Comment (fejta) [inline prow/plugins/dog/dog_test.go]: Seems like shouldError is always false and therefore unnecessary?

File: prow/plugins/dog/dog_test.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	url := "http://localhost"
+	b, err := json.Marshal(&dogResult{
+		URL: url,
+	})
+	blob := string(b)
+	if err != nil {
+		t.Errorf("Failed to encode test data: %v", err)
+	}
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if strings.Contains(r.URL.Path, "bad-json") {
```
Comment (fejta) [inline prow/plugins/dog/dog_test.go]: Malformed json too?

File: prow/plugins/dog/dog_test.go [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+		case err != nil:
+			t.Errorf("%s: unexpected error: %v", tc.name, err)
+		case !re.MatchString(ret):
+			t.Errorf("%s: bad return value: %s", tc.name, ret)
+		}
+	}
+}
+
+// Medium integration test (depends on ability to open a TCP port)
+func TestHttpResponse(t *testing.T) {
```
Comment (fejta) [inline prow/plugins/dog/dog_test.go]: Seems like this wants to have a testcase struct with []byte response, url string, err book } or something like that and then cases such as "good" "http error" "not json" "wrong keys" "not a url"

File: prow/plugins/dog/dog_test.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+func (c fakePack) readDog() (string, error) {
+	return string(c), nil
+}
+
+func TestRealDog(t *testing.T) {
+	if !*human {
+		t.Skip("Real dogs disabled for automation. Manual users can add --human [--category=foo]")
+	}
+	if cat, err := dogURL.readDog(); err != nil {
```
Comment (fejta) [inline prow/plugins/dog/dog_test.go]: cat == dog???

**Comment by fejta** [comment]
/lgtm
/hold

Think we can improve the testing a bit. Otherwise looks great. Can't wait to use it.

File: prow/plugins/dog/dog_test.go (hunk 8 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog_test.go]: done

File: prow/plugins/dog/dog_test.go (hunk 9 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog_test.go]: Refactored these

File: prow/plugins/dog/dog_test.go (hunk 10 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog_test.go]: SGTM, pretty much did just this.

File: prow/plugins/dog/dog_test.go (hunk 11 above)
Comment (BenTheElder) [reply prow/plugins/dog/dog_test.go]: done

**Comment by BenTheElder** [comment]
/hold cancel
improved testing

**Pull request merged**